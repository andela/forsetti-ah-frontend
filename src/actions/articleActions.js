import {
  GET_ARTICLES_SUCCESS,
  GET_ARTICLES_FAIL,
  GET_ARTICLES_BEGIN,
  GET_SINGLE_ARTICLE,
  ARTICLE_NOT_FOUND
} from '../action-types';
import axios from '../config/axiosConfig';

const loadingStateHandler = () => ({
  type: GET_ARTICLES_BEGIN
});

const getAritlcesSuccessHandler = (payload = [], count = 0) => ({
  type: GET_ARTICLES_SUCCESS,
  payload,
  count
});

const getAritlcesFailureHandler = (payload = '') => ({
  type: GET_ARTICLES_FAIL,
  payload
});

const getAritlces = page => async (dispatch) => {
  dispatch(loadingStateHandler());
  try {
    const response = await axios.get(`/article?page=${page}`);
    const { rows, count } = response.data.data.articles;
    return dispatch(getAritlcesSuccessHandler(rows, count));
  } catch (error) {
    return dispatch(getAritlcesFailureHandler(error.message));
  }
};

const singleArticle = (data = {}) => {
  const article = data.data === undefined ? {} : data.data.data[0];
  return {
    type: GET_SINGLE_ARTICLE,
    article
  };
};

const articleNotFound = (error = '') => ({
  type: ARTICLE_NOT_FOUND,
  error,
});

const getSingleArticle = slug => async (dispatch) => {
  try {
    const res = await axios.get(`/article/${slug}`);
    return dispatch(singleArticle(res));
  } catch (error) {
    return dispatch(articleNotFound(error.message));
  }
};

export {
  getSingleArticle,
  singleArticle,
  getAritlces,
  loadingStateHandler,
  getAritlcesSuccessHandler,
  getAritlcesFailureHandler,
  articleNotFound
};
