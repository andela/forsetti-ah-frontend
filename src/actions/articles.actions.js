import { createArticleActionTypes } from '../action-types';
import axiosInstance from '../config/axiosConfig';

const {
  CREATE_ARTICLE_LOADING,
  CREATE_ARTICLE,
  CREATE_ARTICLE_ERROR,
  CREATE_ARTICLE_PUBLISHED
} = createArticleActionTypes;

const setLoading = payload => ({
  type: CREATE_ARTICLE_LOADING,
  payload
});

const createArticleError = errorMessage => ({
  type: CREATE_ARTICLE_ERROR,
  payload: errorMessage
});

const createArticle = articleObject => async (dispatch) => {
  try {
    dispatch(setLoading(true));
    const token = localStorage.getItem('token');
    const {
      tagList, currentSlug, imageChanged, image, published
    } = articleObject;
    const method = currentSlug ? 'edit' : 'post';
    const formData = new FormData();

    Object.keys(articleObject).forEach(key => formData.append(key, articleObject[key]));
    formData.delete('currentSlug');
    formData.delete('tagList');
    formData.delete('imageChanged');
    tagList.forEach((tag) => {
      formData.append('tagList[]', tag);
    });
    if (!imageChanged) {
      formData.delete('image');
    }

    const header = {
      'Content-Type': 'multipart/form-data',
      authorization: `Bearer ${token}`
    };

    if (method === 'post') {
      const { data } = await axiosInstance.post('/article', formData, {
        headers: header
      });
      localStorage.setItem('slug', data.data.article.slug);
      const actionType = published ? CREATE_ARTICLE_PUBLISHED : CREATE_ARTICLE;
      return dispatch({
        type: actionType,
        payload: data.data
      });
    }

    const { data } = await axiosInstance.put(`/article/${currentSlug}`, formData, {
      headers: header
    });
    const actionType = published ? CREATE_ARTICLE_PUBLISHED : CREATE_ARTICLE;
    return dispatch({
      type: actionType,
      payload: data.data
    });
  } catch (err) {
    const { response: { data: { message } } } = err;
    return dispatch(createArticleError(message));
  }
};

export { setLoading, createArticle, createArticleError };
