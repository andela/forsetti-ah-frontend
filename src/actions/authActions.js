import { auth } from '../action-types';
import axios from '../config/axiosConfig';

const { LOGIN_USER_SUCCESS, LOGIN_USER_FAILURE, LOGIN_USER_BEGIN } = auth;

/**
 * Show loading state
 */
export const loadingStateHandler = () => ({
  type: LOGIN_USER_BEGIN
});

/**
 * Handles successful api call to the login API
 * @param {object} response
 */
export const onLoginSuccessHandler = (response = {}, rememberLogin = false) => ({
  type: LOGIN_USER_SUCCESS,
  payload: response,
  rememberLogin
});

/**
 * Handles axios error with api call to the login API
 * @param {object} response
 */
export const onFailureHandler = response => ({
  type: LOGIN_USER_FAILURE,
  payload: response
});

/**
 * @param {object} loginObject email and password object
 * @param {boolean} rememberLogin
 * @returns {function} dispatch functions
 */
const loginUser = (loginObject, rememberLogin) => async (dispatch) => {
  try {
    dispatch(loadingStateHandler());
    const response = await axios.post('/auth/login', loginObject);
    return dispatch(onLoginSuccessHandler(response.data, rememberLogin));
  } catch (error) {
    return dispatch(onFailureHandler(error.response.data));
  }
};

export default loginUser;
