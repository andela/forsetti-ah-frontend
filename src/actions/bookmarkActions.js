import { BOOKMARK_ARTICLE_SUCCESS, BOOKMARK_ARTICLE_FAILURE } from '../action-types';
import axios from '../config/axiosConfig';

const token = window.localStorage.getItem('token');
const bookmarkArticleSuccessHandler = status => ({
  type: BOOKMARK_ARTICLE_SUCCESS,
  status
});
const bookmarkArticleFailureHandler = status => ({
  type: BOOKMARK_ARTICLE_FAILURE,
  status
});


const bookmarkArticle = articleId => async (dispatch) => {
  try {
    const response = await axios.post(`/article/${articleId}/bookmark`, articleId, {
      headers: {
        Authorization: `Bearer ${token}`
      }
    });
    return dispatch(bookmarkArticleSuccessHandler(response.status));
  } catch (error) {
    return dispatch(bookmarkArticleFailureHandler(error, error));
  }
};

export {
  bookmarkArticle,
  bookmarkArticleSuccessHandler,
  bookmarkArticleFailureHandler
};
