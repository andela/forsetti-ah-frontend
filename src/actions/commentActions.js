import {
  COMMENT_LOADING,
  POST_COMMENT_SUCCESS,
  POST_COMMENT_ERROR,
  POST_THREAD_COMMENT
} from '../action-types';
import instance from '../config/axiosConfig';

const token = localStorage.getItem('token') || null;
const commentLoading = payload => ({
  type: COMMENT_LOADING,
  payload
});

const postCommentError = errorMessage => ({
  type: POST_COMMENT_ERROR,
  payload: errorMessage
});

const postComment = commentObject => async (dispatch) => {
  try {
    dispatch(commentLoading(true));
    let { select } = commentObject;
    if (select === 'comment') select = 'normal';
    const requestBody = JSON.stringify({
      comment: commentObject.comment,
      commentType: select
    });

    const { data } = await instance.post(`/article/${commentObject.slug}/comment`, requestBody, {
      headers: {
        'Content-Type': 'application/json',
        authorization: `Bearer ${token}`
      }
    });
    dispatch(commentLoading(false));
    return dispatch({
      type: POST_COMMENT_SUCCESS,
      payload: data.data
    });
  } catch (err) {
    dispatch(commentLoading(false));
    const { response: { data: { message } } } = err;
    return dispatch(postCommentError(message));
  }
};

const postThreadComment = commentObject => async (dispatch) => {
  try {
    const { threadComment, slug, id } = commentObject;
    let { select } = commentObject;
    if (select === 'comment') select = 'normal';

    const requestBody = JSON.stringify({
      comment: threadComment,
      commentType: select
    });

    const { data } = await instance.post(`/article/${slug}/comment/${id}/thread`, requestBody, {
      headers: {
        'Content-Type': 'application/json',
        authorization: `Bearer ${token}`
      }
    });

    return dispatch({
      type: POST_THREAD_COMMENT,
      payload: data.data
    });
  } catch (err) {
    const { response: { data: { message } } } = err;
    return dispatch(postCommentError(message));
  }
};

export {
  postComment,
  postCommentError,
  postThreadComment,
  commentLoading
};
