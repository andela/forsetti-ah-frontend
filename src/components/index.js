import Header from './common/Header';
import Footer from './common/Footer';
import VerticalListItems from './common/VerticalListItems';
import HorizontalListItems from './common/HorizontalListItems';
import Backdrop from './common/Backdrop';
import SideDrawer from './SideDrawer';
import NavBarItems from './common/NavBarItems';
import Button from './common/Button';

export {
  Header,
  Footer,
  VerticalListItems,
  HorizontalListItems,
  Backdrop,
  SideDrawer,
  NavBarItems,
  Button
};
