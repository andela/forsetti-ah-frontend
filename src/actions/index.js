import { showSideDrawerAction, hideSideDrawerAction } from './navBarActions';
import {
  openModalAction,
  closeModalAction,
  openTagsModal,
  closeTagsModal,
  openSignupModalAction,
  closeSignUpModalAction
} from './modalActions';
import {
  loginUser,
  getUserData,
  getUserDataFailure,
  getUserDataSuccess,
  signupBegin,
  signupSuccess,
  signupFailure,
  signUpUser
} from './authActions';
import {
  loadingStateHandler,
  getAritlcesSuccessHandler,
  getAritlcesFailureHandler,
  getAritlces,
  getSingleArticle,
  singleArticle
} from './articleActions';
import { setLoading, createArticle, createArticleError } from './articles.actions';
import { getNotifications, getUserNotifications } from './notifcationActions';
import {
  closeModal,
  openModal,
  forgotPassword,
  resetPassword,
  getToken,
  openResetModal,
  closeResetModal
} from './resetPasswordActions';
import {
  getCurrentProfile,
  setProfileLoading,
  updateProfile
} from './profileActions';

export {
  showSideDrawerAction,
  hideSideDrawerAction,
  loginUser,
  openModalAction,
  closeModalAction,
  openTagsModal,
  closeTagsModal,
  loadingStateHandler,
  getAritlcesSuccessHandler,
  getAritlcesFailureHandler,
  getAritlces,
  getSingleArticle,
  singleArticle,
  setLoading,
  createArticle,
  createArticleError,
  getUserData,
  getUserDataFailure,
  getUserDataSuccess,
  openSignupModalAction,
  closeSignUpModalAction,
  signupBegin,
  signupSuccess,
  signupFailure,
  signUpUser,
  getNotifications,
  getUserNotifications,
  closeModal,
  openModal,
  forgotPassword,
  resetPassword,
  getToken,
  openResetModal,
  closeResetModal,
  getCurrentProfile,
  setProfileLoading,
  updateProfile
};
