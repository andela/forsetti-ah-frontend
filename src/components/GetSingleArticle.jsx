import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { getSingleArticle } from '../actions/articleActions';
import Comments from './Comments';
import ArticleHeader from './ArticleHeader';
import ArticleBody from './ArticleBody';
import Notfound from './NotFound';

export class GetSingleArticle extends Component {
  async componentDidMount() {
    const {
      getOneArticle,
      match: {
        params: {
          slug
        }
      },
      article: {
        title
      }
    } = this.props;
    document.title = title || 'Home to the creative | Authors Haven';
    await getOneArticle(slug);
  }

  render() {
    const {
      article: {
        success
      }
    } = this.props;
    if (!success) {
      return (
        <Notfound text='article' />
      );
    }
    const {
      article: {
        author: {
          image,
          username
        },
        body,
        claps,
        readingTime,
        title,
        image: articleImage,
        createdAt
      }
    } = this.props;

    const headerProps = {
      title,
      image,
      username,
      readingTime,
      articleImage,
      createdAt
    };
    const bodyProps = {
      body, claps
    };
    return (
      <Fragment>
        <ArticleHeader {...headerProps} />
        <div className='container'>
          <ArticleBody {...bodyProps} />
          <Comments />
        </div>
      </Fragment>
    );
  }
}

export const mapStateToProps = state => ({
  article: state.article,
});

export const mapDispatchToProps = {
  getOneArticle: getSingleArticle
};
export const SingleArticle = connect(mapStateToProps, mapDispatchToProps)(withRouter(GetSingleArticle));
