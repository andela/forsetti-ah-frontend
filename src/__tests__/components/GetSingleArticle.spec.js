import React from 'react';
import { shallow } from 'enzyme';
import { GetSingleArticle, mapStateToProps } from '../../components/GetSingleArticle';
import { getSingleArticle } from '../../actions';

const article = {
  article: {
    author: {
      id: 'ffffabd5-4a5b',
      bio: 'test bio',
      image: 'image',
      username: 'john doe',
    },
    id: 'ffffabd5-4a5b',
    body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    claps: '2 claps',
    comments: '0 comments',
    description: 'Lorem ipsum dolor sit amet',
    rating: '0',
    image: null,
    readingTime: '2',
    slug: 'Lorem ipsum dolor sit amet-12345',
    title: 'Lorem ipsum dolor sit amet',
    success: true
  }
};
const error = {
  article: {
    success: false
  }
}

const auth = {
  article: {
    article: ''
  },
  auth: {
    token: 'jhkjkjhjh'
  }
}
describe('Get single article component', () => {
  test('should render get single article component', () => {
    const wrapper = shallow(<GetSingleArticle
      match={{
        params: { slug: 'ffffabd5-4a5b' }
      }}
      getOneArticle={getSingleArticle}
      {...article}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  test('should render 404 page when single article not found', () => {
    const wrapper = shallow(<GetSingleArticle
      match={{
        params: { slug: 'ffffabd5-4a5b' }
      }}
      getOneArticle={getSingleArticle}
      {...error}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  test('should return mapStateToProps object', () => {
    const action = mapStateToProps(auth);
    expect(action).toEqual({
      article: auth.article,
      token: auth.auth.token,
    });
  });
});
