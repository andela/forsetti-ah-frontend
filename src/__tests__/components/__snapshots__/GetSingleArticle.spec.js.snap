// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get single article component should render 404 page when single article not found 1`] = `ShallowWrapper {}`;

exports[`Get single article component should render get single article component 1`] = `ShallowWrapper {}`;
